File: 08/Java/TripleExpression/Negate.java
package TripleExpression;

public class Negate<T> implements TripleExpression<T> {
	private TripleExpression<T> value;
    private Operator<T> op;
    public Negate(TripleExpression<T> val, Operator<T> op) {
        value = val;
        this.op = op;
    }

    public T evaluate(T x, T y, T z) throws Exception {
        T a = value.evaluate(x, y, z);
        return op.negate(a);
    }
}

File: 08/CSharp/TripleExpressionApp/TripleExpressionApp/Negate.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    public class Negate<T> : ITripleExpression<T>
    {
        private ITripleExpression<T> value;
        private IOperator<T> op;
        public Negate(ITripleExpression<T> value, IOperator<T> op)
        {
            this.value = value;
            this.op = op;
        }

        public T Evaluate(T x, T y, T z)
        {
            T a = value.Evaluate(x, y, z);
            return op.Negate(a);
        }
    }
}